### PR TITLE
Accessibility and localization improvements

### PR DESCRIPTION
**Pull request by eduardb**
#### TL;DR

Application was not a11y/l10n compliant, so we made sure it is ^^
#### Problem

Many views were missing a proper content description. Application didn't support RTL locales because it used left/right layout attributes. Also, there was no way to have weather condition strings translated.
#### Solution

Added content descriptions to views where it made sense, extracted translatable strings for weather conditions, and use start/end attributes instead of left/right to better support RTL locales.
##### Test(s) added

 Yes, tested that correct weather description is fetched for various weather condition ids.
##### Screenshots

| Before | After |
| --- | --- |
| device-2016-04-19-180118 | device-2016-04-19-175958 |
| ------ | ----- |
| a11y_before | a11y_after |
###### Paired with

no one

### CODE REVIEW

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java [hunk 1]
```diff
@@ -193,37 +193,46 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
             mFriendlyDateView.setText(friendlyDateText);
             mDateView.setText(dateText);
 
-            // Read description from cursor and update view
-            String description = data.getString(COL_WEATHER_DESC);
+            // Get description from weather condition ID
+            String description = Utility.getStringForWeatherCondition(getActivity(), weatherId);
```
Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:197]: Does fragment not have getString?

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java [hunk 2]
```diff
@@ -193,37 +193,46 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
...
+            // has text describing it in the same UI component
+            mIconView.setContentDescription(getString(R.string.a11y_forecast_icon, description));
 
             // Read high temperature from cursor and update view
             boolean isMetric = Utility.isMetric(getActivity());
 
             double high = data.getDouble(COL_WEATHER_MAX_TEMP);
             String highString = Utility.formatTemperature(getActivity(), high);
             mHighTempView.setText(highString);
+            mHighTempView.setContentDescription(getString(R.string.a11y_high_temp, highString));
```
Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:211]: :100: setting explicit cd

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java (hunk 1 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:197]: It has, but I don't see what difference it makes. There is some logic behind how the weather description is got, and it is extracted in the `Utility` class now.

File: app/build.gradle [hunk 3]
```diff
@@ -2,7 +2,7 @@ apply plugin: 'com.android.application'
 
 android {
     compileSdkVersion 23
-    buildToolsVersion "23.0.2"
+    buildToolsVersion "23.0.3"
```
Comment (blundell) [inline app/build.gradle:5]: nice when this is done as a separate PR. Usually on bigger projects updating the tools can make ppl giving your PR a +1 take a whoooole lot longer, so just safer to split for faster merges

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java (hunk 2 above)
Comment (blundell) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:211]: @ataulm correct me if I am wrong, but if a view has `setText` then this text would be used for the content description any way?

File: app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java [hunk 4]
```diff
@@ -75,17 +75,18 @@ public void bindView(View view, Context context, Cursor cursor) {
         ViewHolder viewHolder = (ViewHolder) view.getTag();
 
         int viewType = getItemViewType(cursor.getPosition());
+        int weatherId = cursor.getInt(ForecastFragment.COL_WEATHER_CONDITION_ID); // TODO bad practice!?!?!
```
Comment (blundell) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java]: what is bad practice?

File: app/src/main/java/com/example/android/sunshine/app/Utility.java [hunk 5]
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
...
+        int stringId;
+        if (weatherId >= 200 && weatherId <= 232) {
+            stringId = R.string.condition_2xx;
+        } else if (weatherId >= 300 && weatherId <= 321) {
+            stringId = R.string.condition_3xx;
+        } else {
+            stringId = context.getResources().getIdentifier(
+                    String.format(Locale.US, "condition_%d", weatherId),
+                    "string",
+                    context.getPackageName());
```
Comment (blundell) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: you could use `BuildConfig.APPLICATION_ID` to avoid the use of context on this line

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -116,4 +116,63 @@
...
+  <string name="cold">Cold</string>
+  <string name="hot">Hot</string>
+  <string name="windy">Windy</string>
+  <string name="hail">Hail</string>
+  <string name="calm">Calm</string>
+  <string name="light_breeze">Light Breeze</string>
+  <string name="gentle_breeze">Gentle Breeze</string>
+  <string name="breeze">Breeze</string>
+  <string name="fresh_breeze">Fresh Breeze</string>
+  <string name="strong_breeze">Strong Breeze</string>
```
Comment (blundell) [inline app/src/main/res/values/strings.xml]: I would recommend using case sensitive names so that you know exactly what it holds i.e.

`<string name="Strong_Breeze">Strong Breeze</string>`

File: build.gradle [hunk 7]
```diff
@@ -5,7 +5,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.1.0-alpha5'
+        classpath 'com.android.tools.build:gradle:2.1.0-beta3'
```
Comment (blundell) [inline build.gradle:8]: 👍 living on the edge

File: app/build.gradle (hunk 3 above)
Comment (eduardb) [inline app/build.gradle:5]: Makes sense, I didn't think about it because this is a small project.

File: app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java (hunk 4 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java]: Oh snap, it was a comment for myself that I committed by mistake. I think it's bad practice that the adapter has knowledge about the fragment in which is being used.

File: build.gradle (hunk 7 above)
Comment (eduardb) [inline build.gradle:8]: Always 😎

File: app/build.gradle (hunk 3 above)
Comment (blundell) [inline app/build.gradle:5]: yeah exactly

File: app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java (hunk 4 above)
Comment (blundell) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java]: more explanatory comments will always help yourself when you come back to the code :-)

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java (hunk 2 above)
Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:211]: You are correct. Sometimes it's desired to set different CD than the text -
for succinctness or to provide more context.

Generally I would recommend making all these not important for
accessibility and putting a content description on the whole card/row.

File: app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java (hunk 4 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java]: I resolved the TODO here: d92e389

File: app/src/main/java/com/example/android/sunshine/app/Utility.java (hunk 5 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: b5b04af

File: app/src/main/res/values/strings.xml (hunk 6 above)
Comment (eduardb) [inline app/src/main/res/values/strings.xml]: 499e714

File: app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -16,6 +16,7 @@
     android:allowBackup="true"
     android:icon="@mipmap/ic_launcher"
     android:label="@string/app_name"
+    android:supportsRtl="true"
```
Comment (xrigau) [inline app/src/main/AndroidManifest.xml:19]: 👍

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java [hunk 9]
```diff
@@ -193,37 +193,46 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
...
             // Read low temperature from cursor and update view
             double low = data.getDouble(COL_WEATHER_MIN_TEMP);
             String lowString = Utility.formatTemperature(getActivity(), low);
             mLowTempView.setText(lowString);
+            mLowTempView.setContentDescription(getString(R.string.a11y_low_temp, lowString));
 
             // Read humidity from cursor and update view
             float humidity = data.getFloat(COL_WEATHER_HUMIDITY);
             mHumidityView.setText(getActivity().getString(R.string.format_humidity, humidity));
+            mHumidityView.setContentDescription(mHumidityView.getText());
```
Comment (xrigau) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java]: isn't this what paul is talking about? this is redundant (same for the 2 below) cc @ataulm

File: build.gradle (hunk 7 above)
Comment (xrigau) [inline build.gradle:8]: beta is less edgy than alpha #justSayin

File: app/src/main/java/com/example/android/sunshine/app/Utility.java [hunk 10]
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
...
+        } else {
+            stringId = context.getResources().getIdentifier(
+                    String.format(Locale.US, "condition_%d", weatherId),
+                    "string",
+                    BuildConfig.APPLICATION_ID);
+
+            if (stringId == 0) {
+                return context.getString(R.string.condition_unknown, weatherId);
+            }
+            return context.getString(stringId);
```
Comment (xrigau) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: unnecessary return

File: app/src/main/java/com/example/android/sunshine/app/Utility.java [hunk 11]
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
...
+            stringId = R.string.condition_2xx;
+        } else if (weatherId >= 300 && weatherId <= 321) {
+            stringId = R.string.condition_3xx;
+        } else {
+            stringId = context.getResources().getIdentifier(
+                    String.format(Locale.US, "condition_%d", weatherId),
+                    "string",
+                    BuildConfig.APPLICATION_ID);
+
+            if (stringId == 0) {
```
Comment (xrigau) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: sounds like an if-else `if string == 0 do this otherwise do ->` https://github.com/eduardb/Sunshine-Version-2/pull/1/files#diff-7537c0d5330829deab3c57d11a53aa5dR280

File: app/src/main/java/com/example/android/sunshine/app/Utility.java [hunk 12]
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
...
 
+    /**
+     * Helper method to provide the string according to the weather
+     * condition id returned by the OpenWeatherMap call.
+     *
+     * @param context   Android context
+     * @param weatherId from OpenWeatherMap API response
+     * @return string for the weather condition. null if no relation is found.
+     */
+    public static String getStringForWeatherCondition(Context context, int weatherId) {
```
Comment (xrigau) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java:271]: this could be unit tested 👅  I see 4 unit tests 💥

**Comment by xrigau** [comment]
good stuff! just a couple of comments but the rest LGTM 👍

File: app/src/main/java/com/example/android/sunshine/app/Utility.java (hunk 11 above)
Comment (xrigau) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: ah sorry I misread that, but this `if` could be moved to line 290 (as in - doesn't need to be a nested `if`)

Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: nested `ifs` can generally be extracted into smaller methods so you have at most one level of indentation per method

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -116,4 +116,63 @@
   <string name="empty_forecast_list_server_error">No weather information available. The server is not returning valid data. Please check for an updated version of Sunshine.</string>
   <string name="empty_forecast_list_invalid_location">No weather information available. The location in settings is not recognized by the weather server.</string>
 
+  <!-- A11y -->
+  <string name="a11y_forecast">Forecast: <xliff:g id="condition">%1$s</xliff:g></string>
```
Comment (ataulm) [inline app/src/main/res/values/strings.xml]: sik xliffin

Comment (ataulm) [inline app/src/main/res/values/strings.xml]: the xliff tag also takes an `example` attribute I think, so you could provide an example of what the token's value to help translators understand the context.

File: build.gradle (hunk 7 above)
Comment (ataulm) [inline build.gradle:8]: hahaha

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java (hunk 9 above)
Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java]: yep this one is redundant as TalkBack would read the TextView's text property anyway.

File: app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java [hunk 14]
```diff
@@ -75,44 +77,46 @@ public void bindView(View view, Context context, Cursor cursor) {
...
+        viewHolder.descriptionView.setContentDescription(context.getString(R.string.a11y_forecast, description));
 
-        // For accessibility, add a content description to the icon field
-        viewHolder.iconView.setContentDescription(description);
-
-        // Read user preference for metric or imperial temperature units
-        boolean isMetric = Utility.isMetric(context);
+        // For accessibility, we don't want a content description for the icon field
+        // because the information is repeated in the description view and the icon
+        // is not individually selectable
```
Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java:109]: I thought I saw a string resource for the icon's content description. if it's not used, should be deleted.

Note, ImageView's do not need content descriptions - the lint check is wrong (or out of date at least)!

An ImageView will be ignored by TalkBack unless you put an explicit content description, add `importantForAccessibility="yes"` or make it focusable/clickable.

ImageButtons on the other hand are buttons and as such are clickable by default, so you need a content description or it needs to be marked as `importantForAccessibiilty="no"` to be ignored. If it's a button that does something and you hide it from TalkBack, then an affordance for TalkBack users must be available somewhere else.

cc @blundell

**Comment by ataulm** [comment]
👍 lgtm, but :x: until you add a gif that shows the effect this had on TalkBack. You can "display text" in TalkBack's developer settings (found at the bottom of TalkBack Settings)

😁

File: app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java (hunk 14 above)
Comment (blundell) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java:109]: what versions of Android do you refer to though?

Comment (ataulm) [inline app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java:109]: 16+ at least but I guess all.

File: app/src/main/java/com/example/android/sunshine/app/Utility.java (hunk 10 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: 85d1bdf

File: app/src/main/java/com/example/android/sunshine/app/Utility.java (hunk 11 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java]: cd2bb74

File: app/src/main/java/com/example/android/sunshine/app/Utility.java (hunk 12 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/Utility.java:271]: Is this what you meant? 3af1eb0

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (eduardb) [inline app/src/main/res/values/strings.xml]: 31ebfe4

File: app/src/main/java/com/example/android/sunshine/app/DetailFragment.java (hunk 9 above)
Comment (eduardb) [inline app/src/main/java/com/example/android/sunshine/app/DetailFragment.java]: 7429b7b

**Comment by eduardb** [comment]
@ataulm I added the gifs, the "display text" is very useful, thanks for the tip 💯

**Comment by ataulm** [comment]
👍 lgetm

you can use the DOWN THEN RIGHT gesture to open the global context menu, and then pressing and holding in the center, drag up to "Read from top".

This'll go through all the elements from the top of your screen (it won't scroll if you have a list though), so you can see the order in which it iterates through your views.

**Comment by eduardb** [comment]
Another day, another thing that I learn from you :D
Thanks guys for the review, I really appreciate it @ataulm @blundell @xrigau
